File: ios/Classes/Utils/BarcodeFormatMapper.swift
// BarcodeFormatMapper.swift
// ScanKit - Modern barcode scanner for Flutter
//
// This file provides mapping between string format identifiers
// and native Vision/VisionKit symbology types.

import Vision
import VisionKit

// MARK: - Barcode Format Mapper

/// Maps between ScanKit format strings and native symbology types.
///
/// Format string identifiers:
/// - 2D codes: "qr", "aztec", "dataMatrix", "pdf417"
/// - 1D product: "ean8", "ean13", "upca", "upce"
/// - 1D industrial: "code39", "code93", "code128", "codabar", "itf"
struct BarcodeFormatMapper {

    // MARK: - Vision Framework (for image scanning)

    /// Map format strings to Vision symbologies.
    /// Used for VNDetectBarcodesRequest in image scanning.
    static func mapFormatsToSymbologies(_ formats: [String]?) -> [VNBarcodeSymbology] {
        guard let formats = formats, !formats.isEmpty else {
            return allSymbologies()
        }

        var symbologies: [VNBarcodeSymbology] = []
        for format in formats {
            if let symbology = formatToSymbology(format) {
                symbologies.append(symbology)
            }
        }
        return symbologies.isEmpty ? [.qr] : symbologies
    }

    /// Convert a format string to Vision symbology.
    static func formatToSymbology(_ format: String) -> VNBarcodeSymbology? {
        switch format {
        case "qr": return .qr
        case "aztec": return .aztec
        case "dataMatrix": return .dataMatrix
        case "pdf417": return .pdf417
        case "ean8": return .ean8
        case "ean13": return .ean13
        case "upca": return .upce  // Note: Vision maps UPC-A to UPC-E
        case "upce": return .upce
        case "code39": return .code39
        case "code93": return .code93
        case "code128": return .code128
        case "codabar":
            if #available(iOS 15.0, *) { return .codabar }
            return nil
        case "itf": return .itf14
        default: return nil
        }
    }

    /// Convert Vision symbology to format string.
    static func symbologyToFormat(_ symbology: VNBarcodeSymbology) -> String {
        switch symbology {
        case .qr: return "qr"
        case .aztec: return "aztec"
        case .dataMatrix: return "dataMatrix"
        case .pdf417: return "pdf417"
        case .ean8: return "ean8"
        case .ean13: return "ean13"
        case .upce: return "upce"
        case .code39, .code39Checksum, .code39FullASCII, .code39FullASCIIChecksum:
            return "code39"
        case .code93, .code93i: return "code93"
        case .code128: return "code128"
        case .itf14: return "itf"
        default:
            if #available(iOS 15.0, *) {
                if symbology == .codabar { return "codabar" }
            }
            return "unknown"
        }
    }

    /// All supported Vision symbologies.
    private static func allSymbologies() -> [VNBarcodeSymbology] {
        var symbologies: [VNBarcodeSymbology] = [
            .qr, .aztec, .dataMatrix, .pdf417,
            .ean8, .ean13, .upce,
            .code39, .code93, .code128, .itf14
        ]
        if #available(iOS 15.0, *) {
            symbologies.append(.codabar)
        }
        return symbologies
    }

    // MARK: - VisionKit Framework (for DataScannerViewController)

    /// Map format strings to VisionKit symbologies.
    /// Used for DataScannerViewController (iOS 16+).
    @available(iOS 16.0, *)
    static func mapFormatsToVisionKitSymbologies(_ formats: [String]?) -> [VNBarcodeSymbology] {
        guard let formats = formats, !formats.isEmpty else {
            return [
                .qr, .aztec, .dataMatrix, .pdf417,
                .ean8, .ean13, .upce,
                .code39, .code93, .code128, .codabar, .itf14
            ]
        }

        var symbologies: [VNBarcodeSymbology] = []
        for format in formats {
            if let symbology = formatToVisionKitSymbology(format) {
                symbologies.append(symbology)
            }
        }
        return symbologies.isEmpty ? [.qr] : symbologies
    }

    /// Convert format string to VisionKit symbology (iOS 16+).
    @available(iOS 16.0, *)
    private static func formatToVisionKitSymbology(_ format: String) -> VNBarcodeSymbology? {
        switch format {
        case "qr": return .qr
        case "aztec": return .aztec
        case "dataMatrix": return .dataMatrix
        case "pdf417": return .pdf417
        case "ean8": return .ean8
        case "ean13": return .ean13
        case "upca": return .upce
        case "upce": return .upce
        case "code39": return .code39
        case "code93": return .code93
        case "code128": return .code128
        case "codabar": return .codabar
        case "itf": return .itf14
        default: return nil
        }
    }
}

File: android/src/main/kotlin/dev/scankit/scankit/utils/BarcodeFormatMapper.kt
/**
 * BarcodeFormatMapper.kt
 * ScanKit - Modern barcode scanner for Flutter
 *
 * This file provides mapping between string format identifiers
 * and ML Kit barcode format constants.
 */
package dev.scankit.scankit.utils

import com.google.mlkit.vision.barcode.common.Barcode

/**
 * Maps between ScanKit format strings and ML Kit barcode format constants.
 *
 * Format string identifiers:
 * - 2D codes: "qr", "aztec", "dataMatrix", "pdf417"
 * - 1D product: "ean8", "ean13", "upca", "upce"
 * - 1D industrial: "code39", "code93", "code128", "codabar", "itf"
 */
object BarcodeFormatMapper {

    /**
     * Convert a format string to ML Kit format constant.
     * @param format The format string identifier
     * @return ML Kit format constant, or null if unknown
     */
    fun formatStringToMlKit(format: String): Int? {
        return when (format) {
            "qr" -> Barcode.FORMAT_QR_CODE
            "aztec" -> Barcode.FORMAT_AZTEC
            "dataMatrix" -> Barcode.FORMAT_DATA_MATRIX
            "pdf417" -> Barcode.FORMAT_PDF417
            "ean8" -> Barcode.FORMAT_EAN_8
            "ean13" -> Barcode.FORMAT_EAN_13
            "upca" -> Barcode.FORMAT_UPC_A
            "upce" -> Barcode.FORMAT_UPC_E
            "code39" -> Barcode.FORMAT_CODE_39
            "code93" -> Barcode.FORMAT_CODE_93
            "code128" -> Barcode.FORMAT_CODE_128
            "codabar" -> Barcode.FORMAT_CODABAR
            "itf" -> Barcode.FORMAT_ITF
            else -> null
        }
    }

    /**
     * Convert ML Kit format constant to format string.
     * @param format ML Kit format constant
     * @return Format string identifier
     */
    fun mlKitFormatToString(format: Int): String {
        return when (format) {
            Barcode.FORMAT_QR_CODE -> "qr"
            Barcode.FORMAT_AZTEC -> "aztec"
            Barcode.FORMAT_DATA_MATRIX -> "dataMatrix"
            Barcode.FORMAT_PDF417 -> "pdf417"
            Barcode.FORMAT_EAN_8 -> "ean8"
            Barcode.FORMAT_EAN_13 -> "ean13"
            Barcode.FORMAT_UPC_A -> "upca"
            Barcode.FORMAT_UPC_E -> "upce"
            Barcode.FORMAT_CODE_39 -> "code39"
            Barcode.FORMAT_CODE_93 -> "code93"
            Barcode.FORMAT_CODE_128 -> "code128"
            Barcode.FORMAT_CODABAR -> "codabar"
            Barcode.FORMAT_ITF -> "itf"
            else -> "unknown"
        }
    }

    /**
     * Convert a list of format strings to ML Kit format constants.
     * @param formats List of format string identifiers, or null for all formats
     * @return List of ML Kit format constants (empty if all formats should be used)
     */
    fun mapFormatsToMlKit(formats: List<String>?): List<Int> {
        if (formats.isNullOrEmpty()) return emptyList()
        return formats.mapNotNull { formatStringToMlKit(it) }
    }
}
